File: javaSpringShip/src/main/java/pl/battleships/javaspringship/api/GameController.java
package pl.battleships.javaspringship.api;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.ExceptionHandler;
import pl.battleships.api.GameApi;
import pl.battleships.api.dto.GameDto;
import pl.battleships.api.dto.PositionDto;
import pl.battleships.api.dto.ShipDto;
import pl.battleships.api.dto.ShotStatusDto;
import pl.battleships.core.exception.*;
import pl.battleships.javaspringship.service.GameService;

import java.util.List;

@Controller
@RequiredArgsConstructor
@Slf4j
public class GameController implements GameApi {

    protected final GameService gameService;

    @Override
    public ResponseEntity<List<ShipDto>> find(String id, Boolean destroyed) {
        return ResponseEntity.ok(gameService.findShips(id, destroyed));
    }

    @Override
    public ResponseEntity<List<PositionDto>> getAllShots(String id) {
        return ResponseEntity.ok(gameService.getAllShots(id));
    }

    @Override
    public ResponseEntity<List<ShipDto>> joinGame(GameDto game) {
        return ResponseEntity.ok(gameService.joinTheGame(game));
    }

    @Override
    public ResponseEntity<ShotStatusDto> shot(String id, PositionDto position) {
        return ResponseEntity.ok(gameService.opponentShot(id, position));
    }

    @ExceptionHandler(DuplicatedGameException.class)
    protected ResponseEntity<Void> duplicatedGame() {
        return ResponseEntity.status(HttpStatus.CONFLICT).build();
    }

    @ExceptionHandler(NoGameFoundException.class)
    protected ResponseEntity<Void> noGameFound() {
        log.error("Game not found");
        return ResponseEntity.status(HttpStatus.NOT_FOUND).build();
    }

    @ExceptionHandler(GameOverException.class)
    protected ResponseEntity<Void> gameOver() {
        log.error("Game over");
        return ResponseEntity.status(HttpStatus.GONE).build();
    }

    @ExceptionHandler(InvalidParamException.class)
    protected ResponseEntity<Void> invalidParams() {
        return ResponseEntity.status(HttpStatus.UNPROCESSABLE_ENTITY).build();
    }

    @ExceptionHandler(InvalidMoveException.class)
    protected ResponseEntity<Void> invalidMoveException(){
        return ResponseEntity.status(HttpStatus.FORBIDDEN).build();
    }

}

File: kotlinSpringShip/src/main/kotlin/pl/battleships/kotlinspringship/api/GameController.kt
package pl.battleships.kotlinspringship.api

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestController
import pl.battleships.api.GameApi
import pl.battleships.api.dto.GameDto
import pl.battleships.api.dto.GameStatusDto
import pl.battleships.api.dto.PositionDto
import pl.battleships.api.dto.ShotStatusDto
import pl.battleships.core.exception.GameOverException
import pl.battleships.kotlinspringship.service.GameService

@RestController
@CrossOrigin
class GameController(private val gameService: GameService) : GameApi {
    override fun getAllShots(id: String): ResponseEntity<List<PositionDto>> {
        return ResponseEntity.ok(gameService.getAllShots(id))
    }

    override fun getGameStatus(id: String): ResponseEntity<GameStatusDto> {
        return ResponseEntity.ok(gameService.getGameStatus(id))
    }

    override fun joinGame(gameDto: GameDto): ResponseEntity<Unit> {
        gameService.joinTheGame(gameDto)
        return ResponseEntity.ok().build()
    }

    override fun shot(id: String, positionDto: PositionDto): ResponseEntity<ShotStatusDto> {
        return ResponseEntity.ok(gameService.opponentShot(id, positionDto))
    }

    @ExceptionHandler(GameOverException::class)
    protected fun gameOver(): ResponseEntity<Void?>? {
        return ResponseEntity.status(HttpStatus.GONE).build()
    }

}